API Documentation

```diff
@@ -0,0 +1,740 @@
+
+### **`step_type` function**
+
+```rust
+pub fn step_type(&mut self, name: &str) -> StepTypeHandler
+
+```
+
+**Description:**
+
+Adds a new step type with the specified name to the circuit and returns a **`StepTypeHandler`** instance.
+
+**Arguments:**
+
+- **`name: &str`** - The name of the new step type.
+
+**Return:**
+
+- **`StepTypeHandler`** - A **`StepTypeHandler`** instance representing the added step type.
+
+### **`step_type_def` function**
+
+```rust
+pub fn step_type_def<D>(&mut self, handler: StepTypeHandler, def: D)
+where
+    D: FnOnce(&mut StepTypeContext<F, StepArgs>),
+
+```
+
+**Description:**
+
+Defines a step type using the provided **`StepTypeHandler`** and a function that takes a mutable reference to a **`StepTypeContext`**. This function typically adds constraints to a step type and defines witness generation.
+
+**Arguments:**
+
+- **`handler: StepTypeHandler`** - The **`StepTypeHandler`** instance representing the step type to define.
+- **`def: D`** - The function that defines the step type by taking a mutable reference to a **`StepTypeContext`**.
+
+### **`trace` function**
+
+```rust
+pub fn trace<D>(&mut self, def: D)
+where
+    D: Fn(&mut dyn TraceContext<StepArgs>, TraceArgs) + 'static,
+
+```
+
+**Description:**
+
+Sets the trace function for the circuit. The trace function is responsible for adding step instantiations defined in step_type_def.
+
+**Arguments:**
+
+- **`def: D`** - The trace function, which takes a mutable reference to a **`TraceContext`** and a **`TraceArgs`** instance.
+
+### **fixed_gen**
+
+```rust
+pub fn fixed_gen<D>(&mut self, def: D)
+where
+    D: Fn(&mut dyn FixedGenContext<F>) + 'static,
+
+```
+
+**Description:**
+
+Sets the fixed generation function for the circuit. The fixed generation function is responsible for assigning fixed values to fixed column **`Queriable`**. It is entirely left for the user to implement and is Turing complete. Users typically generate cell values and call the `**assign**` function to fill the fixed columns.
+
+**Arguments:**
+
+- **`def: D`** - The fixed generation function, which takes a mutable reference to a **`FixedGenContext`**. See more information about this trait and its assign function below.
+
+### `FixedGenContext` trait and `assign` function (non-DSL)
+
+**`FixedGenContext`** is a trait that represents a fixed column generation context. It provides an interface for assigning values to fixed columns in a circuit at the specified offset.
+
+```rust
+pub trait FixedGenContext<F> {
+    fn assign(&mut self, offset: usize, lhs: Queriable<F>, rhs: F);
+}
+
+```
+
+The assign function takes a **`Queriable`** object representing the fixed column (lhs) and the value (rhs) to be assigned.
+
+### **`pragma_first_step` function**
+
+```rust
+pub fn pragma_first_step(&mut self, step_type: StepTypeHandler)
+
+```
+
+**Description:**
+
+Enforce the type of the first step for the circuit using the **`StepTypeHandler`** of the desired StepType
```

Comment:
I would say that enforce by adding a constraint to the circuit

Reply:
Edits for pragma_first_step:
Enforce the type of the first step by adding a constraint to the circuit.

for pragma_last_step:
Enforce the type of the last step by adding a constraint to the circuit.